File: validation/upstream/openai-agents-python/ase_agent.py
"""Run a real upstream OpenAI Agents flow and emit ASE adapter events."""

from __future__ import annotations

import argparse
import os
import sys
from pathlib import Path
from typing import TYPE_CHECKING

ROOT = Path(__file__).resolve().parents[3]
UPSTREAM_ROOT = ROOT / ".upstream" / "openai-agents-python"
if str(UPSTREAM_ROOT) not in sys.path:
    sys.path.insert(0, str(UPSTREAM_ROOT))

if TYPE_CHECKING:
    from agents import Agent

    from ase.adapters.frameworks import OpenAIAgentsAdapter

AGENT_ID = "openai-refund-agent"
SESSION_ID = "sess-openai-example"
SPAN_ID = "openai-refund-agent:issue_refund"
TARGET_URL = "https://api.example.com/refunds"


def _parse_args() -> argparse.Namespace:
    """Accept an optional event output path for certification runs."""
    default_path = Path(
        os.environ.get(
            "ASE_ADAPTER_EVENT_SOURCE",
            str(Path(__file__).with_name("events.generated.jsonl")),
        )
    )
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--events-out",
        type=Path,
        default=default_path,
    )
    return parser.parse_args()


def _build_agent(adapter: OpenAIAgentsAdapter) -> Agent:
    """Create a deterministic OpenAI Agents run around one tool call."""
    from agents import Agent, function_tool
    from tests.fake_model import FakeModel
    from tests.test_responses import (
        get_final_output_message,
        get_function_tool_call,
    )

    @function_tool
    def issue_refund(order_id: str) -> str:
        adapter.tool_start(
            AGENT_ID,
            span_id=SPAN_ID,
            tool_kind="http_api",
            method="POST",
            target=TARGET_URL,
            name="issue_refund",
            data={"order_id": order_id},
        )
        adapter.tool_end(
            AGENT_ID,
            span_id=SPAN_ID,
            tool_kind="http_api",
            method="POST",
            target=TARGET_URL,
            data={"status_code": 200, "order_id": order_id},
        )
        return f"refunded {order_id}"

    initial_output = [get_function_tool_call("issue_refund", '{"order_id":"ord-001"}')]
    model = FakeModel(initial_output=initial_output)
    model.set_next_output([get_final_output_message("refunded ord-001")])
    return Agent(name=AGENT_ID, model=model, tools=[issue_refund])


def main() -> int:
    """Execute a real upstream OpenAI Agents run for ASE validation."""
    from agents import Runner

    from ase.adapters.frameworks import OpenAIAgentsAdapter
    from ase.adapters.io import JsonlFileEventSink

    args = _parse_args()
    args.events_out.unlink(missing_ok=True)
    adapter = OpenAIAgentsAdapter(JsonlFileEventSink(args.events_out))
    agent = _build_agent(adapter)
    status = "passed"
    adapter.agent_start(AGENT_ID, name=AGENT_ID)
    adapter.session_write(AGENT_ID, SESSION_ID, key="intent", value="refund")
    try:
        result = Runner.run_sync(agent, input="refund ord-001")
        print(result.final_output)
        print(args.events_out)
        return 0
    except BaseException:
        status = "failed"
        raise
    finally:
        adapter.agent_end(AGENT_ID, status=status)


if __name__ == "__main__":
    raise SystemExit(main())

File: validation/upstream/openai-agents-js/ase_agent.ts
import { appendFile, mkdir, rm } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { z } from 'zod';

import {
  Agent,
  run,
  setDefaultModelProvider,
  tool,
  Usage,
} from '../../../.upstream/openai-agents-js/packages/agents-core/src/index.ts';
import type { ModelResponse } from '../../../.upstream/openai-agents-js/packages/agents-core/src/model.ts';
import type {
  AssistantMessageItem,
  FunctionCallItem,
} from '../../../.upstream/openai-agents-js/packages/agents-core/src/types/protocol.ts';
import {
  FakeModel,
  FakeModelProvider,
} from '../../../.upstream/openai-agents-js/packages/agents-core/test/stubs.ts';

const AGENT_ID = 'openai-agents-js-refund-agent';
const SPAN_ID = 'openai-agents-js-refund-agent:issue_refund';
const TARGET_URL = 'https://api.example.com/refunds';
const DEFAULT_EVENT_PATH = resolve(
  dirname(fileURLToPath(import.meta.url)),
  'events.generated.jsonl',
);
const EVENT_PATH = resolve(
  process.env.ASE_ADAPTER_EVENT_SOURCE ?? DEFAULT_EVENT_PATH,
);

type AdapterEvent = {
  protocol_version: number;
  event_type: string;
  event_id: string;
  timestamp_ms: number;
  span_id?: string;
  agent_id?: string;
  name?: string;
  tool_kind?: string;
  method?: string;
  target?: string;
  status?: string;
  message?: string;
  data?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
};

let nextEventId = 0;

function metadata(): Record<string, unknown> {
  return {
    adapter_name: 'openai-agents-js',
    framework: 'openai-agents-js',
    language: 'typescript',
    transport: 'jsonl-stdio',
  };
}

function functionToolCall(): FunctionCallItem {
  return {
    id: 'fc_1',
    type: 'function_call',
    name: 'issue_refund',
    callId: 'call_1',
    status: 'completed',
    arguments: '{"order_id":"ord-001"}',
    providerData: {},
  };
}

function finalMessage(): AssistantMessageItem {
  return {
    type: 'message',
    role: 'assistant',
    status: 'completed',
    content: [
      {
        type: 'output_text',
        text: 'refunded ord-001',
        providerData: { annotations: [] },
      },
    ],
    providerData: {},
  };
}

async function emit(event: AdapterEvent): Promise<void> {
  await appendFile(EVENT_PATH, `${JSON.stringify(event)}\n`, 'utf-8');
}

async function emitEvent(
  eventType: string,
  extra: Omit<
    AdapterEvent,
    'protocol_version' | 'event_type' | 'event_id' | 'timestamp_ms'
  >,
): Promise<void> {
  nextEventId += 1;
  await emit({
    protocol_version: 1,
    event_type: eventType,
    event_id: `evt_${nextEventId}`,
    timestamp_ms: Date.now(),
    metadata: metadata(),
    ...extra,
  });
}

function buildAgent(): Agent {
  const modelResponses: ModelResponse[] = [
    { output: [functionToolCall()], usage: new Usage() },
    { output: [finalMessage()], usage: new Usage() },
  ];
  const model = new FakeModel(modelResponses);
  const issueRefundTool = tool({
    name: 'issue_refund',
    description: 'Issue one deterministic refund',
    parameters: z.object({ order_id: z.string() }),
    execute: async ({ order_id }: { order_id: string }) => {
      await emitEvent('tool_start', {
        agent_id: AGENT_ID,
        span_id: SPAN_ID,
        tool_kind: 'http_api',
        method: 'POST',
        target: TARGET_URL,
        name: 'issue_refund',
        data: { order_id },
      });
      await emitEvent('tool_end', {
        agent_id: AGENT_ID,
        span_id: SPAN_ID,
        tool_kind: 'http_api',
        method: 'POST',
        target: TARGET_URL,
        status: 'passed',
        data: { order_id, status_code: 200 },
      });
      return `refunded ${order_id}`;
    },
  });
  return new Agent({
    name: AGENT_ID,
    instructions: 'Issue a refund with one tool call.',
    model,
    tools: [issueRefundTool],
  });
}

async function main(): Promise<number> {
  await mkdir(dirname(EVENT_PATH), { recursive: true });
  await rm(EVENT_PATH, { force: true });
  setDefaultModelProvider(new FakeModelProvider());
  const agent = buildAgent();
  let status = 'passed';
  await emitEvent('agent_start', { agent_id: AGENT_ID, name: AGENT_ID });
  try {
    const result = await run(agent, 'refund ord-001');
    console.log(result.finalOutput);
    console.log(EVENT_PATH);
    return 0;
  } catch (error) {
    status = 'failed';
    throw error;
  } finally {
    await emitEvent('agent_end', {
      agent_id: AGENT_ID,
      status,
      message: status === 'passed' ? undefined : 'agent run failed',
    });
  }
}

main().then(
  (code) => process.exit(code),
  (error: unknown) => {
    console.error(error);
    process.exit(1);
  },
);

